gax/src/fallbackServiceStub.ts

```diff
@@ -194,11 +207,55 @@ export function generateServiceStub(
                     }
                     streamArrayParser.emit('error', err);
                   } else if (callback) {
-                    callback(err);
+                    // This supports a legacy Apiary behavior that allows
+                    // empty 204 responses. If we do not intercept this potential error
+                    // from decodeResponse in fallbackRest
+                    // it will cause libraries to erroneously throw an
+                    // error when the call succeeded. This error cannot be checked in
+                    // fallbackRest.ts because decodeResponse does not have the necessary
+                    // context about the response to validate the status code + ok-ness
+                    if (!response204Ok) {
+                      callback(err);
+                      return; // Explicitly return otherwise TS will ha
+                    } else {
+                      // format the empty response the same way we format non-empty responses in fallbackRest.ts
+                      const emptyMessage = serializer.fromProto3JSON(
+                        rpc.resolvedResponseType!,
+                        JSON.parse('{}'),
+                      );
+                      const resp = rpc.resolvedResponseType!.toObject(
+                        emptyMessage!,
+                        defaultToObjectOptions,
+                      );
+                      callback(null, resp);
+                      return; // Explicitly return otherwise TS will have issues
+                    }
                   } else {
-                    throw err;
+                    // This supports a legacy Apiary behavior that allows
+                    // empty 204 responses. If we do not intercept this potential error
+                    // from decodeResponse in fallbackRest
+                    // it will cause libraries to erroneously throw an
+                    // error when the call succeeded. This error cannot be checked in
+                    // fallbackRest.ts because decodeResponse does not have the necessary
+                    // context about the response to validate the status code + ok-ness
+                    if (!response204Ok) {
+                      throw err;
+                    }
+                    // format the empty response the same way we format non-empty responses in fallbackRest.ts
+                    const emptyMessage = serializer.fromProto3JSON(
+                      rpc.resolvedResponseType!,
+                      JSON.parse('{}'),
+                    );
+                    const resp = rpc.resolvedResponseType!.toObject(
+                      emptyMessage!,
+                      defaultToObjectOptions,
+                    );
+                    return resp;
                   }
                 }
+                // If we reach here, it's a cancelled AbortError
+                // return to make sure TS doesn't get mad
+                return;
```

Comment:
assuming our existing test coverage covers this case (an error that isn't 204 & empty)

Reply:
I will check on this, I totally forgot I was the one that added this

Reply:
this is not necessary - we do have tests for cancel, and I think with the restructure where we no longer have the "if callback" logic, TS doesn't care about this path/it's encompassed elsewhere